ca/ca_test.go

```diff
@@ -904,7 +910,49 @@ func TestIssueCertificateForPrecertificate(t *testing.T) {
 	test.Assert(t, list, "returned cert doesn't contain SCT list")
 }
 
+// dupeSA returns a non-error to GetCertificate in order to simulate a request
+// to issue a final certificate with a duplicate serial.
+type dupeSA struct {
+	mockSA
+}
+
+func (m *dupeSA) GetCertificate(ctx context.Context, serial string) (core.Certificate, error) {
+	return core.Certificate{}, nil
+}
+
+func TestIssueCertificateForPrecertificateDuplicateSerial(t *testing.T) {
+	testCtx := setup(t)
+	sa := &mockSA{}
+	ca, err := NewCertificateAuthorityImpl(
+		testCtx.caConfig,
+		sa,
+		testCtx.pa,
+		testCtx.fc,
+		testCtx.stats,
+		testCtx.issuers,
+		testCtx.keyPolicy,
+		testCtx.logger,
+		nil)
+	test.AssertNotError(t, err, "Failed to create CA")
+
+	orderID := int64(0)
+	issueReq := caPB.IssueCertificateRequest{Csr: CNandSANCSR, RegistrationID: &arbitraryRegID, OrderID: &orderID}
+	precert, err := ca.IssuePrecertificate(ctx, &issueReq)
+	test.AssertNotError(t, err, "Failed to issue precert")
+	_, err = ca.IssueCertificateForPrecertificate(ctx, &caPB.IssueCertificateForPrecertificateRequest{
+		DER:            precert.DER,
+		SCTs:           nil,
+		RegistrationID: &arbitraryRegID,
+		OrderID:        new(int64),
+	})
+	if err == nil {
+		t.Fatal("Expected error issuing duplicate serial but got none.")
+	}
```

Comment:
Does the err generated in this case have any distinguishing features to test for here? e.g. type of error or message?

Reply:
Good call! Added a string test. Thanks.